corehq/apps/user_importer/helpers.py

```diff
@@ -9,3 +12,68 @@ def spec_value_to_boolean_or_none(user_spec_dict, key):
         return value
     else:
         return None
+
+
+class UserChangeLogger(object):
+    def __init__(self, domain, user, is_new_user, changed_by_user, changed_via):
+        self.domain = domain
+        self.user = user
+        self.is_new_user = is_new_user
+        self.changed_by_user = changed_by_user
+        self.changed_via = changed_via
+
+        self.original_user_doc = self.user.to_json()
+        self.fields_changed = {}
+        self.messages = []
+
+        self._save = False  # flag to check if log needs to be saved for updates
+
+    def add_changes(self, changes):
+        # ignore for new user since the whole user doc is logged for a new user
+        if self.is_new_user:
+            return
+        for name, new_value in changes.items():
+            if self.original_user_doc[name] != new_value:
+                self.fields_changed[name] = new_value
+                self._save = True
+
+    def add_change_message(self, message):
+        # ignore for new user since the whole user doc is logged for a new user
+        if self.is_new_user:
+            return
+        self.messages.append(message)
+        self._save = True
+
+    def add_info(self, info):
+        # useful info for display like names
+        self.messages.append(info)
+        self._save = True
+
+    def save(self):
+        self._include_user_data_changes()
+        if self.is_new_user:
+            log_user_change(
+                self.domain,
+                self.user,
+                changed_by_user=self.changed_by_user,
+                changed_via=self.changed_via,
+                message=". ".join(self.messages),
+                action=UserModelAction.CREATE,
+            )
+        else:
+            if not self._save:
+                return
+            log_user_change(
+                self.domain,
+                self.user,
+                changed_by_user=self.changed_by_user,
+                changed_via=self.changed_via,
+                message=". ".join(self.messages),
+                fields_changed=self.fields_changed
+            )
```

Comment:
Could be simplified:
```suggestion
        if self.is_new_user or self._save:
            action = UserModelAction.CREATE if self.is_new_user else UserModelAction.UPDATE
            fields_changed = None if self.is_new_user else self.fields_changed
            log_user_change(
                self.domain,
                self.user,
                changed_by_user=self.changed_by_user,
                changed_via=self.changed_via,
                message=". ".join(self.messages),
                fields_changed=fields_changed,
                action=action,
            )
```

Reply:
https://github.com/dimagi/commcare-hq/pull/29935/commits/dc98af9ada3e684523c90bbca5e2d0fcccf098cf